### PR TITLE
DNN: load fp16 ONNX model as fp32

### PR DESCRIPTION
**Pull request by zihaomu**
[Related extra data](https://github.com/opencv/opencv_extra/pull/993)
[Download link](https://github.com/zihaomu/zihaomu/files/9393786/mobilenetv2_fp16_v7.tar.gz)

The FP16 model is half the size of the FP32 model, which will be edge device friendly.
Currently, OpenCV `ONNX_importer` cannot load FP16 models correctly and DNN also doesn't have full FP16 backend support.
So, my idea is that in the first step we make OpenCV DNN load FP16 ONNX model as FP32 model, and every layer will be computed as FP32.

Here is a [FP16 text recognition model(CRNN)](https://github.com/opencv/opencv_zoo/pull/79) which was converted by my GSoC student Yiyao @Charles-258, and we found it even has better accuracy performance than the original FP32 model.
And I have tested these FP16 CRNN models in the existing High-Level API, `testTextRecognitionModel`, and it works well. Maybe we can use this to replace the original FP32 model.

### How to convert ONNX fP32 model to FP16 model?
```py
from onnx import load_model, save_model
from onnxmltools.utils import float16_converter

dir = "MODEL_DIR_PATH"
onnx_model = load_model(os.path.join(dir, "resnet50v1.onnx"))
trans_model = float16_converter.convert_float_to_float16(onnx_model, keep_io_types=False)
save_model(trans_model, os.path.join(dir, "resnet50v1_FP16.onnx"))

```

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Review by vpisarev — APPROVED** [review]

**Comment by vpisarev** [comment]
the PR is important and is ready to be merged

**Review by rogday — COMMENTED** [review]
I think we should log this behavior so that users are not surprised that OpenCV DNN takes twice as much RAM as other frameworks, otherwise looks good.

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 1]
```diff
@@ -767,6 +767,25 @@ Mat getMatFromTensor(const opencv_onnx::TensorProto& tensor_proto)
             Mat(sizes, CV_32FC1, val).copyTo(blob);
         }
     }
+    else if (datatype == opencv_onnx::TensorProto_DataType_FLOAT16)
+    {
+        // FIXME, for now, we only load FP16 Tensor as FP32 Mat, full support for FP16 is required in the future.
+        static bool warningOnce = true;
```
Comment (rogday) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:773]: If the end user decides to import two networks, it'll see only one warning. I think it's okay to spam this on every converted tensor during import.

Comment (alalek) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:773]: BTW, OpenCV has `CV_LOG_ONCE_WARNING` macro.

Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:773]: Hi @rogday. A model may output hundreds of lines of warning, which is too messy. I think one line of warning is enough to let the user know what's going on inside.

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 2]
```diff
@@ -767,6 +767,25 @@ Mat getMatFromTensor(const opencv_onnx::TensorProto& tensor_proto)
...
+            warningOnce = false;
+        }
+
+        if (!tensor_proto.float_data().empty()) {
+            const ::google::protobuf::RepeatedField<float> field = tensor_proto.float_data();
+            Mat(sizes, CV_32FC1, (void*)field.data()).copyTo(blob);
+        }
+        else {
+            char* val = const_cast<char*>(tensor_proto.raw_data().c_str());
+            Mat(sizes, CV_16FC1, val).convertTo(blob, CV_32FC1);
```
Comment (alalek) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:786]: > raw_data()

We have several reports about alignment issues from ARM platforms. E.g. here: #20725
Mat of CV_16F expects 2 bytes alignment.

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp (hunk 1 above)
Comment (rogday) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:773]: Then it should be one line per model import imo.

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp (hunk 2 above)
Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:786]: Thanks for code reviewing. I will try to test this under the ARMv7 platform.

Comment (fengyuentau) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:786]: Can this problem be fixed in another pull request? We need this feature for GSoC projects.

**Review by fengyuentau — APPROVED** [review]
👍

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp (hunk 2 above)
Comment (asmorkalov) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:780]: @zihaomu @alalek According my research the behavior is different for different ARM generations:
- ARM v8 supports unaligned load and `vld1_s16` does not case SIGBUG, but has some performance penalty.
- ARM v7 has different behavior that depends on SoC, kernel and compiler flags and may throw SIGBUS. The behavior is managed by `/proc/cpu/alignment`.

So, I propose to ensure element-size alignment to be in safe.

Comment (asmorkalov) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:780]: https://stackoverflow.com/questions/16548059/how-to-trap-unaligned-memory-access

Comment (zihaomu) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:780]: @asmorkalov Thanks for the detailed investigation. I will try to fix this.

**Comment by zihaomu** [comment]
Hi @alalek and @asmorkalov, can you please check if the update can fix the data alignment issue?

**Comment by asmorkalov** [comment]
@zihaomu I checked other tickets and OpenCV patches that are alignment related and found general solution for it that is managed by macro CV_STRONG_ALIGNMENT. See https://github.com/opencv/opencv/blob/bb71cb200eafbe54ed4dfd31ad47a2d8de6838a4/modules/core/include/opencv2/core/cvdef.h#L351.

I propose to use it and rework the patch in the same way as it's done for int64 tensors in the same function (Macro check + AutoBuffer).
Also related issue: https://github.com/opencv/opencv/issues/16373

**Comment by asmorkalov** [comment]
Also it looks like fp64 branch suffers from the same problem, but it's not popular and the issue is invisible.

**Comment by zihaomu** [comment]
Great thanks! @asmorkalov I will follow your suggestion and re-implement it with `CV_STRONG_ALIGNMENT `.

**Review by asmorkalov — APPROVED** [review]

**Comment by asmorkalov** [comment]
```
[  FAILED  ] Test_ONNX_nets.MobileNet_v2_FP16/0, where GetParam() = OCV/OCL
[  FAILED  ] Test_ONNX_nets.MobileNet_v2_FP16/1, where GetParam() = OCV/OCL_FP16
[  FAILED  ] Test_ONNX_nets.MobileNet_v2_FP16/2, where GetParam() = OCV/CPU
```

**Comment by zihaomu** [comment]
Thanks for the reminder, @asmorkalov. there seems to be a depth bug on float16 and I'm trying to fix it.

**Comment by asenyaev** [comment]
@zihaomu sure, I'll check it.

**Comment by asenyaev** [comment]
@zihaomu, the model was downloaded manually on Windows machine and the pipeline for Windows passed.

**Comment by zihaomu** [comment]
Hi, `ONNX_importer.cpp` may encounter bugs when load float16 model.
When the format version of ONNX is lower than 7, `dnn/misc/opencv_onnx.pb.cc` will parse float 16 tensor into int32, which will cause the weight to be all 0.
Since the `dnn/misc/opencv_onnx.pb.cc` is generated by `opencv-onnx.proto` which was taken from `https://github.com/onnx/onnx/blob/main/onnx/onnx.proto`, I think this bug is not our responsibility.

In my test, this path works fine with model format version >= 7.

**Review by alalek — APPROVED** [review]
LGTM :+1:

**Review by asmorkalov — APPROVED** [review]

**Review by rogday — APPROVED** [review]
:+1:

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp [hunk 3]
```diff
@@ -771,10 +782,25 @@ Mat getMatFromTensor(const opencv_onnx::TensorProto& tensor_proto)
...
         CV_LOG_ONCE_WARNING(NULL, "DNN: load FP16 model as FP32 model, and it takes twice the FP16 RAM requirement.");
-        if (!tensor_proto.float_data().empty())
+
+        // ONNX saves float 16 data in two format: int32 and raw_data.
+        // Link: https://github.com/onnx/onnx/issues/4460#issuecomment-1224373746
+        if (!tensor_proto.int32_data().empty())
         {
-            const ::google::protobuf::RepeatedField<float> field = tensor_proto.float_data();
-            Mat(sizes, CV_32FC1, (void*)field.data()).copyTo(blob);
+            const int offset = isLittleEndianCPU() ? 0 : 1;
```
Comment (fengyuentau) [inline modules/dnn/src/onnx/onnx_graph_simplifier.cpp:790]: IIRC, we have this check in CMake scripts. Can we use a precomp macro here?

File: modules/dnn/test/test_onnx_importer.cpp [hunk 4]
```diff
@@ -16,6 +16,52 @@ static std::string _tf(TString filename, bool required = true)
 {
     return findDataFile(std::string("dnn/onnx/") + filename, required);
 }
+        void printShape(InputArray blob_)
```
Comment (fengyuentau) [inline modules/dnn/test/test_onnx_importer.cpp:19]: unnecessary commit

File: modules/dnn/test/test_onnx_importer.cpp [hunk 5]
```diff
@@ -16,6 +16,52 @@ static std::string _tf(TString filename, bool required = true)
...
+            Mat blob = blob_.getMat();
+            auto shapeA = shape(blob);
+            for (int i = 0; i < shapeA.size(); i++)
+            {
+                std::cout<<"x"<<shapeA[i];
+            }
+            std::cout<<std::endl;
+        }
+
+void printblob(InputArray blob_) {
```
Comment (fengyuentau) [inline modules/dnn/test/test_onnx_importer.cpp:30]: unnecessary commit

File: modules/dnn/test/test_onnx_importer.cpp [hunk 6]
```diff
@@ -66,6 +112,11 @@ class Test_ONNX_layers : public DNNTestLayer
             net.setInput(inps[i], inputNames[i]);
         Mat out = net.forward("");
 
+        std::cout<<"out = "<<std::endl;
+        printblob(out);
+        std::cout<<"ref = "<<std::endl;
+        printblob(ref);
```
Comment (fengyuentau) [inline modules/dnn/test/test_onnx_importer.cpp:118]: unnecessary commit

File: modules/dnn/test/test_onnx_importer.cpp [hunk 7]
```diff
@@ -2095,7 +2146,18 @@ TEST_P(Test_ONNX_nets, CNN_MNIST)
 TEST_P(Test_ONNX_nets, MobileNet_v2)
 {
     // output range: [-166; 317], after Softmax [0; 1]
-    testONNXModels("mobilenetv2", pb, default_l1, default_lInf, true);
+//    testONNXModels("mobilenetv2", pb, default_l1, default_lInf, true);
+
+    String rootPath = "/Users/zihaomu/work/opencv_dev/data_test/models/";
```
Comment (fengyuentau) [inline modules/dnn/test/test_onnx_importer.cpp:2151]: unnecessary commit

File: modules/dnn/test/test_onnx_importer.cpp [hunk 8]
```diff
@@ -2095,7 +2146,18 @@ TEST_P(Test_ONNX_nets, CNN_MNIST)
...
+//    Net net = readNet("/Users/zihaomu/work/opencv_dev/data_test/models/CRNN_EN_FP16_sim.onnx");
+//    Net net = readNet(rootPath + "yolov4_fp16.onnx");
+    Net net = readNet(rootPath + "resnet50_fp16.onnx");
```
Comment (fengyuentau) [inline modules/dnn/test/test_onnx_importer.cpp:2154]: Put this in the PR in opencv_extra?

File: modules/dnn/test/test_onnx_importer.cpp (hunk 4 above)
Comment (zihaomu) [reply modules/dnn/test/test_onnx_importer.cpp]: Thx! Fixed, wrong commit.

**Comment by zihaomu** [comment]
Hi, with the help of ONNX teams, we can load all the ONNX float16 models now!
ONNX saves float 16 data to two different way: `int32_t` and `raw_data`.
[ONNX convert float to float16](https://github.com/microsoft/onnxconverter-common/blob/master/onnxconverter_common/float16.py#L61-L73)

**Review by vpisarev — APPROVED** [review]

File: modules/dnn/src/onnx/onnx_graph_simplifier.cpp (hunk 3 above)
Comment (alalek) [reply modules/dnn/src/onnx/onnx_graph_simplifier.cpp:790]: Please reuse `WORDS_BIGENDIAN` macro (from cvconfig.h)

**Pull request merged**